py/desispec/scripts/proc.py

```diff
@@ -979,22 +1006,43 @@ def main(args=None, comm=None):
                 else :
                     log.warning("No SKYCORR file, do you need to update DESI_SPECTRO_CALIB?")
             cmd += " --fit-offsets"
-            runcmd(cmd, inputs=[framefile, fiberflatfile], outputs=[skyfile,])
+
+            err = runcmd(cmd, inputs=[framefile, fiberflatfile], outputs=[skyfile,])
+            if err != 0:
+                error_count += 1
 
             #- sframe = flatfielded sky-subtracted but not flux calibrated frame
             #- Note: this re-reads and re-does steps previously done for picking
             #- sky fibers; desi_proc is about human efficiency,
             #- not I/O or CPU efficiency...
             sframefile = desispec.io.findfile('sframe', args.night, args.expid, camera)
             if not os.path.exists(sframefile):
-                frame = desispec.io.read_frame(framefile)
-                fiberflat = desispec.io.read_fiberflat(fiberflatfile)
-                sky = desispec.io.read_sky(skyfile)
-                apply_fiberflat(frame, fiberflat)
-                subtract_sky(frame, sky, apply_throughput_correction=True)
-                frame.meta['IN_SKY'] = shorten_filename(skyfile)
-                frame.meta['FIBERFLT'] = shorten_filename(fiberflatfile)
-                desispec.io.write_frame(sframefile, frame)
+                missing_inputs = False
+                for filename in [framefile, fiberflatfile, skyfile]:
+                    if not os.path.exists(filename):
+                        log.error(f'Camera {camera} missing sframe input {filename}')
+                        missing_inputs = True
+
+                if missing_inputs:
+                    log.error(f'Camera {camera} missing sframe inputs; skipping')
+                    error_count += 1
+                else:
+                    try:
+                        frame = desispec.io.read_frame(framefile)
+                        fiberflat = desispec.io.read_fiberflat(fiberflatfile)
+                        sky = desispec.io.read_sky(skyfile)
+                        apply_fiberflat(frame, fiberflat)
+                        subtract_sky(frame, sky, apply_throughput_correction=True)
+                        frame.meta['IN_SKY'] = shorten_filename(skyfile)
+                        frame.meta['FIBERFLT'] = shorten_filename(fiberflatfile)
+                        desispec.io.write_frame(sframefile, frame)
+                    except Exception as err:
+                        import traceback
+                        lines = traceback.format_exception(*sys.exc_info())
+                        log.error(f"Camera {camera} sframe raised an exception:")
+                        print("".join(lines))
+                        log.warning(f'Continuing without {sframefile}')
```

Comment:
Are we sure we want to continue if the file isn't written out?

Reply:
Since the sframe file isn't used as input to any downstream step in desi_proc, it isn't necessarily fatal to those steps and they could proceed.  At the same time, error_count is incremented so that the job at the end will register a failure and downstream *jobs* will be blocked.  Even if we did want to stop hard, we still have to let the process continue so that this failing rank can catch up with others that may have succeeded.  Otherwise we're in the situation that prompted this PR: one rank failed and stopped, while the others continued, and then at some downstream MPI barrier/bcast/gather the job got stuck waiting for the failed rank to catch up.

This is similar to the other steps wrapped by runcmd: if it fails, that is logged but the script proceeds and other downstream steps might fail due to missing inputs, but it still lets the N>>1 MPI ranks continue and get to the end of the script and let that decide the ultimate success/fail of the job.

Arguably this warning is a bit to casual about continuing, but it is proceeded by actual error messages.

Reply:
Thank you that makes complete sense.